### PR TITLE
[Outlook](SSO) Move and update Outlook SSO sample to this repo

### PR DESCRIPTION
**Pull request by davidchesnut**
|        Q        |                    A                    |
| --------------- | --------------------------------------- |
| Bug fix?        | no                                |
| New feature?    | no                              |
| New sample?     | no                                |
| Related issues? | none |

## What's in this Pull Request?

This PR moves the Outlook SSO sample into this PnP repo (from: https://github.com/OfficeDev/Outlook-Add-in-SSO)

It also updates the Outlook SSO sample to use latest MSAL and also just has general code cleanup work.

### CODE REVIEW

**Comment by davidchesnut** [comment]
Would be great to have folks follow the readme and ensure this sample is working as expected! Thanks!

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+> 1. Rename the existing file to `MailAppVersionOverridesV1_1.old`.
+> 1. Move the version of the file from this repository into the folder.
+
+You can run the sample right from Visual Studio. Select the **AttachmentDemo** project in **Solution Explorer**, then choose the **Start Action** value you want (under **Add-in** in the properties window). You can choose any installed browser to launch Outlook on the web, or you can choose **Office Desktop Client** to launch Outlook. If you choose **Office Desktop Client**, be sure to configure Outlook to connect to the Office 365 or Outlook.com user you want to install the add-in for.
+
+Press **F5** to build and debug the project. You should be prompted for a user account and password. Be sure to use a user in your Office 365 tenant, or an Outlook.com account. The add-in will be installed for that user, and either Outlook on the web or Outlook will open. Select any message, and you should see the add-in buttons on the Outlook ribbon.
```
Comment (Rick-Kirkham) [inline Samples/auth/Outlook-Add-in-SSO/README.md:149]:
```suggestion
Press **F5** to build and debug the project. You should be prompted for a user account and password. Be sure to use a user in your Office 365 tenant, or an Outlook.com account. The add-in will be installed for that user, and either Outlook on the web or Outlook will open. Select any message **that has one or more attachments**, and you should see the add-in buttons on the Outlook ribbon.
```

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Functions/FunctionFile.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+Office.initialize = function () {
+
+}
+
+async function saveAllAttachments(event) {
+    var attachmentIds = [];
+    let accessToken = await OfficeRuntime.auth.getAccessToken({ allowSignInPrompt: true, allowConsentPrompt: true, forMSGraphAccess: true });
```
Comment (Rick-Kirkham) [inline Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Functions/FunctionFile.js:10]: There's no error handling for the call to getAccessToken and no branching to the fallback method when the call fails.

**Review by Rick-Kirkham — CHANGES_REQUESTED** [review]
Biggest issue is that there's no proper error handling for the call of getAccessToken and no branching to the fallback method. 
Also, the files don't save. 
See my specific comments.

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    - **User consent display name**: Office can act as you.
+    - **User consent description**: Enable Office to call the add-in's web APIs with the same rights that you have.
+
+1. Ensure that **State** is set to **Enabled**.
+
+1. Select **Add scope** .
+
+1. In the **Authorized client applications** section, you identify the applications that you want to authorize to your add-in's web application. Each of the following IDs needs to be pre-authorized.
+
+    - `d3590ed6-52b3-4102-aeff-aad2292ab01c` (Microsoft Office)
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/README.md:73]: added.

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+1. Navigate to the [Azure portal - App registrations](https://go.microsoft.com/fwlink/?linkid=2083908) page to register your app.
+
+1. Sign in with the ***admin*** credentials to your Microsoft 365 tenancy. For example, MyName@contoso.onmicrosoft.com.
+
+1. Select **New registration**. On the **Register an application** page, set the values as follows.
+
+    * Set **Name** to `AttachmentsDemo`.
+    * Set **Supported account types** to **Accounts in any organizational directory (Any Azure AD directory - Multitenant) and personal Microsoft accounts (e.g. Skype, Xbox)**.
+    * In the **Redirect URI** section, ensure that **Web** is selected in the drop down and then set the URI to `https://localhost:44355/AzureADAuth/Authorize`.
+        > [!NOTE]
```
[the remark this answers is not in the document]
Comment (Rick-Kirkham) [reply Samples/auth/Outlook-Add-in-SSO/README.md]: Still needs to be fixed.

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+## Configuring the Sample
+
+Before you run the sample, you'll need to do a few things to make it work properly.
+
+Edit [AttachmentDemo.xml](AttachmentDemo/AttachmentDemoManifest/AttachmentDemo.xml) and replace the `[Enter the Client Id (Application ID obtained from the Azure portal)]` value with the application ID you generated as part of the app registration process.
+
+> [!NOTE]
+> Make sure that the port number in the `Resource` element matches the port used by your project. It should also match the port you used when registering the application.
+
+Edit [Web.config](AttachmentDemoWeb/Web.config) and replace the `[Enter the Client Id (Application ID obtained from the Azure portal)]` value in with the application ID and `[Copy the client secret added to the app from the Azure portal]` with the application password you generated as part of the app registration process.
```
[the remark this answers is not in the document]
Comment (Rick-Kirkham) [reply Samples/auth/Outlook-Add-in-SSO/README.md]: Still needs to be fixed.

**Review by Rick-Kirkham — CHANGES_REQUESTED** [review]

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+1. Move the version of the file from this repository into the folder.
+
+**Add-in in Outlook on desktop**
+
+![The add-in buttons on the ribbon in Outlook on the desktop](buttons-outlook.PNG)
+
+**Add-in in Outlook on the web**
+
+![The add-in buttons in Outlook on the web](buttons-owa.PNG)
+
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/README.md:215]: I added this info to the readme.

[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/README.md:215]: Ok, added steps and screen shots.

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+Before you run the sample, you'll need to do a few things to make it work properly.
+
+1. In Visual Studio, open the **AttachmentDemo.sln** solution file for this sample.
+
+1. In the **Solution Explorer** open **AttachmentDemo > AttachmentDemoManifest > AttachmentDemo.xml**. Then replace the `[Enter the Client Id (Application ID obtained from the Azure portal)]` value, in both places where it appears, with the application ID you generated as part of the app registration process.
+
+    **Note:** Make sure that the port number in the `Resource` element matches the port used by your project. It should also match the port you used when registering the application.
+
+1. In the **Solution Explorer** open **AttachmentDemoWeb > Web.config** and replace the `[Enter the Client Id (Application ID obtained from the Azure portal)]` value with the application ID and `[Copy the client secret added to the app from the Azure portal]` with the client secret you generated as part of the app registration process.
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/README.md]: fixed

File: Samples/auth/Outlook-Add-in-SSO/README.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+1. Navigate to the [Azure portal - App registrations](https://go.microsoft.com/fwlink/?linkid=2083908) page to register your app.
+
+1. Sign in with the ***admin*** credentials to your Microsoft 365 tenancy. For example, MyName@contoso.onmicrosoft.com.
+
+1. Select **New registration**. On the **Register an application** page, set the values as follows.
+
+    * Set **Name** to `AttachmentsDemo`.
+    * Set **Supported account types** to **Accounts in any organizational directory (Any Azure AD directory - Multitenant) and personal Microsoft accounts (e.g. Skype, Xbox)**.
+    * In the **Redirect URI** section, ensure that **Web** is selected in the drop down and then set the URI to `https://localhost:44355/AzureADAuth/Authorize`.
+        **Note:** The port number in the redirect URI (`44355`) may be different on your development machine. You can find the correct port number for your machine by selecting the **AttachmentDemoWeb** project in **Solution Explorer**, then looking at the **SSL URL** setting in the properties window.
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/README.md]: fixed.

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Controllers/AzureADAuthController.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+			{
+				ViewBag.Message = "Please set your client ID and client secret in the Web.config file";
+				return View();
+			}
+
+			ConfidentialClientApplicationBuilder clientBuilder = ConfidentialClientApplicationBuilder.Create(Settings.AzureADClientId).WithClientSecret(Settings.AzureADClientSecret);
+			ConfidentialClientApplication clientApp = (ConfidentialClientApplication)clientBuilder.Build();
+
+			// Generate the parameterized URL for Azure login.
+			string[] graphScopes = { "https://graph.microsoft.com/Files.Read.All", "https://graph.microsoft.com/User.Read" };
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Controllers/AzureADAuthController.cs]: fixed to use the correct scopes

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Functions/FunctionFile.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Functions/FunctionFile.js]: removed this from the sample.

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Controllers/SaveAttachmentsController.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+            var cca = ConfidentialClientApplicationBuilder.Create(ConfigurationManager.AppSettings["ida:ClientID"])
+                                                          .WithRedirectUri(ConfigurationManager.AppSettings["ida:Domain"])
+                                                          .WithClientSecret(ConfigurationManager.AppSettings["ida:Password"])
+                                                          .WithAuthority(ConfigurationManager.AppSettings["ida:Authority"])
+                                                          .Build();
+
+            // MSAL.NET adds the profile, offline_access, and openid scopes itself. It will throw an error if you add
+            // them redundantly here.
+            string[] graphScopes = { "https://graph.microsoft.com/Files.Read.All" };
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Controllers/SaveAttachmentsController.cs:45]: fixed to request correct scopes

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Scripts/MessageRead.js [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    function processMessage(arg) {
+
+        console.log("Message received in processMessage: " + JSON.stringify(arg));
+        let message = JSON.parse(arg.message);
+
+        if (message.status === "success") {
+            // We now have a valid access token.
+            loginDialog.close();
+            let attachmentIds = getSelectedAttachments();
+            saveAttachmentsWithSSO("/api/saveAttachmentsFallback", message.accessToken, attachmentIds);
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Scripts/MessageRead.js]: did some refactoring to make this all clearer.

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Scripts/MessageRead.js [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+            if (exception.code) {
+                handleClientSideErrors(exception);
+            }
+            else {
+                showNotification("EXCEPTION: ", JSON.stringify(exception));
+            }
+        }
+    }
+
+    function saveAttachmentsWithSSO(relativeURL, accessToken, attachmentIds) {
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Scripts/MessageRead.js]: fixed.

File: Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Scripts/MessageRead.js [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+        //Set default SSO options if they are not provided
+        if (options === undefined) options = { allowSignInPrompt: true, allowConsentPrompt: true, forMSGraphAccess: true };
+        
+        showSpinner();
+
+        // Attempt to get an SSO token
+        try {
+            let bootstrapToken = await OfficeRuntime.auth.getAccessToken(options);
+
+            // The /api/saveAttachmentsToOneDrive controller will make the token exchange and use the
```
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/AttachmentDemoWeb/Scripts/MessageRead.js]: fixed.

File: Samples/auth/Outlook-Add-in-SSO/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/auth/Outlook-Add-in-SSO/README.md]: fixed.

**Review by Rick-Kirkham — APPROVED** [review]

**Pull request merged**